### PR TITLE
Add a script to fetch and diff GitHub workflow logs

### PR DESCRIPTION
**Pull request by krcroft**
Adds a stand-alone script `./scripts/diff-workflow-against-master.sh` that does a couple things:
1. Pulls the artifacts and logs for the local branch's latest CI run
2. Diffs any failed jobs with the most recent successful equivalent master branch job

The script is fire-and-forget; it takes no argument and just grinds away for a couple seconds until done.  Even if you're not interested in diffing, it's a convenient way to slurp down our current zip files, reports, and logs. 

It currently constructs a directory in `/tmp` to hold content, and the script will clean away older files too so it doesn't get overrun.

The log diff needs improving though.  @dreamer , if you have something smarter than can take in two "raw logs" and diff-out the meaningful gcc differences (and drop everthing elsee), then we can drop that function (or python) in to improve it.  Right now I'm just stripping the prefix time-stamp and then diffing the raw logs; so the resulting diff file is still pretty cluttered.

None the less, it's a first cut that's working - and we can refine it as we go.  Let me know what you think!

Here's output from the start of the run:

### CODE REVIEW

**Comment by krcroft** [comment]
The script will step you through how to generate a personal auth token.  

I _think_ anyone with a GitHub account should be able do this - so anyone curious out there, please kick the tires and report back!

File: scripts/diff-workflow-against-master.sh [hunk 1]
```diff
@@ -0,0 +1,471 @@
...
+##
+#  Gets one or more return_key(s) from records that have
+#  matching search_key and search_value hits
+#
+function query() {
+	local container="$1"
+	local search_key="$2"
+	local search_value="$3"
+	local return_keys="$4"
+	jq -r --arg value "$search_value"\
```
Comment (dreamer) [inline scripts/diff-workflow-against-master.sh]: `./scripts/diff-workflow-against-master.sh: line 274: jq: command not found`

Add a dependency check on script start; if a user has no `jq` installed, the script should immediately exit with an error (before a user has a chance to generating a token).

File: scripts/diff-workflow-against-master.sh [hunk 2]
```diff
@@ -0,0 +1,471 @@
...
+#
+function init_netrc() {
+	# Check and setup the GitHub personal access token
+	local netrc="$HOME/.netrc"
+	if grep -q "^machine $authority" "$netrc"; then
+		echo "Found credentials for $authority in $netrc"
+		return
+	fi
+	# Get the username
+	clear
```
Comment (dreamer) [inline scripts/diff-workflow-against-master.sh]: Don't clear screen, please.

File: scripts/diff-workflow-against-master.sh [hunk 3]
```diff
@@ -0,0 +1,471 @@
...
+						echo -e "        - diffed    ${yellow}$diff_file${reset}"
+					fi
+				done # jobs_array
+			done # conclusion types
+			echo "  $joiner"
+		done # branch types
+	done # workflows
+}
+
+main
```
Comment (dreamer) [inline scripts/diff-workflow-against-master.sh]: Use `main "$@"`; we don't handle any parameters yet, but let's prepare for it.

File: scripts/diff-workflow-against-master.sh [hunk 4]
```diff
@@ -0,0 +1,471 @@
...
+	parent="/tmp/$repo_postfix-workflows-$USER"
+	assets_dir="$parent/assets"
+	cache_dir="$parent/cache"
+	diffs_dir="$parent/diffs"
+	logs_dir="$parent/logs"
+	declare -gr assets_dir
+	declare -gr cache_dir
+	declare -gr diffs_dir
+	declare -gr logs_dir
+	echo "Initializing storage area: $parent"
```
Comment (dreamer) [inline scripts/diff-workflow-against-master.sh]: It took me a while to figure out where the logs were actually downloaded (I missed this line when reading output) - can you remove this echo from here, place it just before program exits (so it'll be the last line of output) and give make the message less technical, e.g.: "Copy of logs in: <path>".

File: scripts/diff-workflow-against-master.sh (hunk 1 above)
Comment (dreamer) [reply scripts/diff-workflow-against-master.sh]: I know that token is saved before this happens, but some user might get lost and generate a new one.

**Comment by krcroft** [comment]
> store token in the repo in .git/config file instead of ~/.netrc

The token is equivalent to a password, which means its storage point should be secured by default. Git's tools create `.git/config` with default system permissions (typically 644):
```
[kcroft@hoya dosbox-staging] (kc/issue-diff-1 $) ll .git/config 
-rw-r--r-- 1 kcroft users 1.7K Feb 12 08:06 .git/config
```
Unlike SSH's tools, that create sensitive files with restrictive permissions:
```
[kcroft@hoya dosbox-staging] (kc/issue-diff-1 $) ll -d ~/.ssh ~/.ssh/id_ed25519
drwx------ 2 kcroft users 4.0K Nov 28 10:35 /home/kcroft/.ssh
-rw------- 1 kcroft users  411 Aug  4  2017 /home/kcroft/.ssh/id_ed25519
```

Similarly, `netrc` is dedicated to storing credentials and is is the recommended place to store sensitive Curl credentials: https://community.apigee.com/articles/39911/do-you-use-curl-stop-using-u-please-use-curl-n-and.html

We could make the script lock down the `.git` directory and its `config` file -- but that doesn't solve external assumption that `.git/config` is just a regular repo file.  

So some users might zip-up their repo and send it to colleagues; or might be backed up into an IT-manager server farm dedicated to GIT repositories having lower security access requirements than where they back-up secret files.. (mentioning this because some companies give workers some hours/week for personal projects). 

> pre-fill email field with git config user.email

For users who have enabled "Keep my email addresses private" in their user settings, then their `config` will be set with the no-reply address as well, such as `email = krcroft@users.noreply.github.com` (and that won't work when generating a token). But I can certainly try to pull it from here first :), and if it happens to be a noreply address then fallback to asking the user.

File: scripts/diff-workflow-against-master.sh (hunk 4 above)
Comment (krcroft) [reply scripts/diff-workflow-against-master.sh]: Will do; much more convenient!

File: scripts/diff-workflow-against-master.sh (hunk 3 above)
Comment (krcroft) [reply scripts/diff-workflow-against-master.sh]: Makes sense; will do.

File: scripts/diff-workflow-against-master.sh (hunk 2 above)
Comment (krcroft) [reply scripts/diff-workflow-against-master.sh]: It felt a bit more instructive, but I'm 50-50 on how it turned out. A real  TUI would have been cool with something like `dialog` :-)  But yeah, will drop the `clear`s!

File: scripts/diff-workflow-against-master.sh (hunk 1 above)
Comment (krcroft) [reply scripts/diff-workflow-against-master.sh]: Huge catch here; I'm baffled why I didn't do a requirements check first - thank you @dreamer !
Will run a check for `unzip` and any other non-core-util commands I might be using.

**Comment by dreamer** [comment]
> > store token in the repo in .git/config file instead of ~/.netrc
> 
> The token is equivalent to a password, which means its storage point should be secured by default. Git's tools create `.git/config` with default system permissions (typically 644):
> 
> ```
> [kcroft@hoya dosbox-staging] (kc/issue-diff-1 $) ll .git/config 
> -rw-r--r-- 1 kcroft users 1.7K Feb 12 08:06 .git/config
> ```
> 
> Unlike SSH's tools, that create sensitive files with restrictive permissions:

`.git/config` is just a user's file, we can safely change mode on it; but I wo't press the issue (I just dislike placing new config files directly in my home dir).

> ```
> [kcroft@hoya dosbox-staging] (kc/issue-diff-1 $) ll -d ~/.ssh ~/.ssh/id_ed25519
> drwx------ 2 kcroft users 4.0K Nov 28 10:35 /home/kcroft/.ssh
> -rw------- 1 kcroft users  411 Aug  4  2017 /home/kcroft/.ssh/id_ed25519
> ```
> 
> Similarly, `netrc` is dedicated to storing credentials and is is the recommended place to store sensitive Curl credentials: https://community.apigee.com/articles/39911/do-you-use-curl-stop-using-u-please-use-curl-n-and.html
> 
> We could make the script lock down the `.git` directory and its `config` file -- but that doesn't solve external assumption that `.git/config` is just a regular repo file.
> 
> So some users might zip-up their repo and send it to colleagues; or might be backed up into an IT-manager server farm dedicated to GIT repositories having lower security access requirements than where they back-up secret files.. (mentioning this because some companies give workers some hours/week for personal projects).

Yeah, that would be bad… Users are not supposed to do that (receiving zipped git repo is not secure for receiver), but some people might try it.

> > pre-fill email field with git config user.email
> 
> For users who have enabled "**Keep my email addresses private**" in their user settings, then their `config` will be set with the no-reply address as well, such as `email = krcroft@users.noreply.github.com` (and that won't work when generating a token). But I can certainly try to pull it from here first :), and if it happens to be a noreply address then fallback to asking the user.

You misunderstood my suggestion: I think we can safely pre-fill input to email (not skip it altogether), so when it matches user can just press enter.

`read -r -p "mail: " -e -i "foo@bar.baz" variable_name`

mail address will be placed as suggested value, user can press enter or edit it :)

**Comment by krcroft** [comment]
Just came across https://git-scm.com/docs/git-credential-store probably the best option; then use an environment variable to pass the token to curl, so it doesn't show up on the command line.

**Comment by dreamer** [comment]
> Just came across https://git-scm.com/docs/git-credential-store probably the best option; then use an environment variable to pass the token to curl, so it doesn't show up on the command line.

Nice! I didn't know it exists :) (I don't feel we need to implement it during this PR - your call).

File: scripts/diff-workflow-against-master.sh (hunk 1 above)
Comment (krcroft) [reply scripts/diff-workflow-against-master.sh]: Added (for all tools)

File: scripts/diff-workflow-against-master.sh (hunk 2 above)
Comment (krcroft) [reply scripts/diff-workflow-against-master.sh]: Done!

File: scripts/diff-workflow-against-master.sh (hunk 4 above)
Comment (krcroft) [reply scripts/diff-workflow-against-master.sh]: Added.

**Comment by krcroft** [comment]
All set for another round:
 - dependencies (like `jq`) checked
 - switched to git credentials
 - pre-populate email, but only if it's a non-noreply address
 - print storage directory at the end
 - args passed to main
 - removed the `clear`s during setup
 - a bit more verbose when `ctrl-c` is pressed

**Comment by krcroft** [comment]
Suggestion for round two PR: trial your last suggestions: 

>  easy way to create diffs - at least for compilation logs would be to run count-warnings.py -lf <file> on specific logs and then run diff on outputs of warning counter; instead of diff we could even run merge.tool/merge.guitool automatically (description of these options in man git-config).

**Comment by dreamer** [comment]
GitHub interface got confused after I merged this change in manually. It's now available on master branch 675d8b99bf5148d406c591b630c64861cb02d466

@krcroft Thank you! :)

**Pull request closed without merging**